virttest/qemu_vm.py

```diff
@@ -472,19 +472,26 @@ def add_human_monitor(devices, monitor_name, filename):
             cmd += _add_option("mode", "readline")
             return cmd
 
-        def add_qmp_monitor(devices, monitor_name, filename):
+        def add_qmp_monitor(devices, monitor_name, monitor_conn, monitor_tcp):
             if not devices.has_option("qmp"):
                 logging.warn("Fallback to human monitor since qmp is"
                              " unsupported")
-                return add_human_monitor(devices, monitor_name, filename)
+                return add_human_monitor(devices, monitor_name, monitor_conn)
 
             if not devices.has_option("chardev"):
-                return " -qmp unix:'%s',server,nowait" % filename
+                if monitor_tcp == "yes":
+                    return " -qmp unix:host=127.0.0.1,port='%s',server,nowait"\
+                           % monitor_conn
+                return " -qmp unix:'%s',server,nowait" % monitor_conn
 
             monitor_id = "qmp_id_%s" % monitor_name
             cmd = " -chardev socket"
             cmd += _add_option("id", monitor_id)
-            cmd += _add_option("path", filename)
+            if monitor_tcp == "yes":
+                cmd += _add_option("host", "127.0.0.1")
+                cmd += _add_option("port", monitor_conn)
```

Comment:
Here is the old design which use a str to add chardev device.  Now, please try to use `devices.chardev_define_by_params` to add chardev device.
something like:
```
chardev_id = "chardev_%s" % monitor_name    --> use qmp_id_$monitor_name is ok for me too
chardev_params = params.object_params(monitor_name)
if chardev_params.get('chardev_backend') == 'tcp_socket': 
    host = chardev_params.get('chardev_host', '127.0.0.1')
    port = utils_misc.find_free_ports(5000, 6000, 1, host)[0]
    chardev_params['chardev_host'] = host 
    chardev_params['chardev_port'] = str(port)
chardev = devices.chardev_define_by_params(chardev_id, chardev_params, filename)
devices.insert(chardev)
....
cmd = " -mon chardev=%s" % chardev_id
cmd += _add_option("mode", "control")
```

update base.cfg with:
```
monitors = qmpmonitor1
chardev_backend_qmpmonitor1 = unix_socket
```

Reply:
Thanks and update.